sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -497,19 +499,25 @@ case class HashAggregateExec(
       val resultVars = bindReferences[Expression](
         resultExpressions,
         inputAttrs).map(_.genCode(ctx))
+      val evaluateResultVars = evaluateVariables(resultVars)
       s"""
        $evaluateKeyVars
        $evaluateResultBufferVars
+       $evaluateResultVars
        ${consume(ctx, resultVars)}
        """
     } else {
       // generate result based on grouping key
       ctx.INPUT_ROW = keyTerm
       ctx.currentVars = null
-      val eval = bindReferences[Expression](
+      val resultVars = bindReferences[Expression](
         resultExpressions,
         groupingAttributes).map(_.genCode(ctx))
-      consume(ctx, eval)
+      val evaluateResultVars = evaluateVariables(resultVars)
+      s"""
+       $evaluateResultVars
```

Comment:
For non broadcast join cases, the change will force evaluation unnecessarily too. We should move evaluation out of the loop in broadcast join, if possible.

Reply:
What I a bit concern about is; is it semantically ok to defer the evaluation of nondeterministic exprs if HashAggregateExec has these exprs?

I think, to fix this issue, its ok to modify code in the join side if we could find a simpler solution there with no performance regression. But, I have just a question about the design regardless of this issue.

Reply:
oh.. Kris answered my question.. in https://github.com/apache/spark/pull/23731#pullrequestreview-202522824